### PR TITLE
Improve registry storage assertions

### LINKED ISSUE #9685: ID collisions on storage insertion should always panic
The logic in `Storage::insert` panics if an attempt is made to insert an object with the same (id, epoch) pair, but it silently replaces an existing item having the same ID but different epoch.

https://github.com/gfx-rs/wgpu/blob/bb3ab6f796f049230b77f49d3dc1ef03577cb0a3/wgpu-core/src/storage.rs#L87-L92

We can think of no reason to allow this. In cases like https://bugzilla.mozilla.org/show_bug.cgi?id=2014639 and https://bugzilla.mozilla.org/show_bug.cgi?id=2022806, we do eventually / intermittently detect an error, but rejecting ID collision upon insertion is just as easy, and is likely closer to the actual problem.

### PR DESCRIPTION
**Pull request by andyleiserson**
In debugging a Firefox bug, I noticed that `Storage::insert` rejects attempts to insert the same resource ID twice, but silently drops `(index, epoch1)` if asked to insert `(index, epoch2)`. This PR disallows that, and logs more information about expected / found IDs in several of the assertions.

Fixes #9685.

**Testing**
Claude wrote me some tests to sanity-check what the messages looked like, which I've included. They seem a little silly as permanent tests, on the other hand, I did add one for the "silently replace object in different epoch" case that was missed before.

**Squash or Rebase?** Rebase

### CODE REVIEW

File: wgpu-core/src/storage.rs [hunk 1]
```diff
@@ -132,13 +130,86 @@ where
...
+                panic!("Cannot get non-existent resource {id:?}");
+            }
         };
         assert_eq!(
-            epoch, storage_epoch,
-            "{}[{:?}] is no longer alive",
-            self.kind, id
+            epoch,
+            storage_epoch,
+            "Cannot get {id:?}, found other resource {other:?}",
```
Comment (ErichDonGubler) [inline wgpu-core/src/storage.rs:140]: nitpick(non-blocking): I wonder if it would be helpful to mention that it's the epoch that doesn't match?

This feedback applies in other places, too.

File: wgpu-core/src/id.rs [hunk 2]
```diff
@@ -280,14 +261,18 @@ where
...
+}
 
 // This allows `()` to be used as a marker type for tests.
 //
 // We don't want these in production code, since they essentially remove type
 // safety, like how identifiers across different types can be compared.
 #[cfg(test)]
-impl Marker for () {}
+impl Marker for () {
+    const TYPE: &'static str = "Untyped";
```
Comment (ErichDonGubler) [inline wgpu-core/src/id.rs:274]: nitpick(non-blocking): This is an edge-case that doesn't seem significant, but I wonder how confusing this would be with a macro-implemented `Untyped` type. 🙃

File: wgpu-core/src/storage.rs (hunk 1 above)
Comment (ErichDonGubler) [reply wgpu-core/src/storage.rs:140]: praise: It'll be nice to be able to see the other resource's full ID in diagnostic output!

**Review by ErichDonGubler — APPROVED** [review]
lezduit

File: wgpu-core/src/storage.rs (hunk 1 above)
Comment (andyleiserson) [reply wgpu-core/src/storage.rs:140]: > I wonder if it would be helpful to mention that it's the epoch that doesn't match?

I considered that, but I came down on the side of brevity since it seems like these errors will be used more by developers (e.g. crash reports) than by users (where clear, actionable messages matter). (Or hopefully, they'll be used by no one, but that's not been the case so far.)

I'm not opposed to adding more detail if you think it's valuable.

Side gripe, the thing that is more frustrating/confusing for me is that the epoch is stored in the upper 32 bits, but when printed, the order is (index, epoch).
